readme.rst
Sierpinski's Carpet Generation
==============================

This is a fun little script was created as a solution to a
`problem on the dailyprogrammer subreddit community <https://www.reddit.com/r/dailyprogrammer/comments/748ba7/20171004_challenge_334_intermediate_carpet/>`_
.

To run
------

It's a good practice to use virtualenvs to isolate package requirements. This script uses the following
packages:

 - attrs==17.2.0
 - Pillow==4.3.0

**Skip the following if you understand virtualenvs**

Assuming you have python you'll need to do the following. If you don't have python installed there are plenty of
resources, or you can send me a message on twitter.

1. Create a virtualenv. This can be done via ``python -m venv nameofyourvirtualenv``
2. Create an environment variable to get to bin or scripts directory in the virtualenv
3. Upgrade pip in said virtualenv

**Resume reading if you are skipping**

To run the scripts you simply invoke it with the python executable in your venv. No parameters needed. Just watch out when
you create a bmp. The script is still simplistic in that it just writes to the same file, so your **new images will
overwrite your old ones**.



changes.rst
Changes
=======
v0.03
-----
- Introduction of pillow and the creation of color bmp representations of the 
  carpets.
v0.02
-----
- Modular design
- Better UI and rule system
v0.01
-----
- Inital version. Generates carpets as text arrays. Can't change rules at runtime.


sierpinski_0.1.py
#-------------------------------------------------------------------------------
# Sierpinski's carpet
# Version: 0.1
# Date: 2017-10-25
# By: D. Clark
# Description:
# --------------------------------------
# End result of first attempt to create a script to generate a sierpinski's
# carpet. It's crude in that you can't change the number of colors, and it's 
# very brittle.
#-------------------------------------------------------------------------------
rules = {0:'000010000',
        1:'111111111'}

def main():
    raw_parameters = input('Enter number of colors and iterations deliminated by a space:')
    num_colors, num_iterations = raw_parameters.split(' ')

    #initialze the output so that it is the first character of the "0" rule
    output = rules[0][0] + '\n'
    lcv = 0
    while lcv < int(num_iterations)-1:
        next_output = ''
        staged_line_1 = ''
        staged_line_2 = ''
        staged_line_3 = ''
        for focus in output:
            if focus == '\n':
                #End of line prep for next line
                next_output = next_output + '\n'.join([staged_line_1, staged_line_2, staged_line_3]) + '\n'
                staged_line_1 = ''
                staged_line_2 = ''
                staged_line_3 = ''
            else:
                staged_line_1 = staged_line_1 + rules[int(focus)][0:3]
                staged_line_2 = staged_line_2 + rules[int(focus)][3:6]
                staged_line_3 = staged_line_3 + rules[int(focus)][6:9]
        output = next_output
        lcv += 1
    print('***********output***********')
    print(output)

    #a way to store rules
    #how to render

if __name__ == '__main__':
    main()


#Notes
#=====
# * Patterns are generated using string replacement
# * Need to generate 3 lines at once


sierpinski_0.2.py
import attr
#-------------------------------------------------------------------------------
# Sierpinski's carpet
# Version: 0.2
# Date: 2017-10-25
# By: D. Clark
# Description:
# --------------------------------------
# Simplistic utility to generate Sierpinski carpets. It's a command line driven
# script that will take basic commands to generate the patterns.
#
# Todos:
# --------------------------------------
# * Testing
# * Logging
# * Image generation
# * Investigate usage of matrices 
#-------------------------------------------------------------------------------
# Some sane defaults. Need a better place for this Using globals for now.
default_rules = {
    0:'000010000',
    1:'111111111'}
default_iterations = 3
rules = None


@attr.s
class SierpinskiConfig(object):
    output = attr.ib(default='')
    rules = attr.ib(default=default_rules)
    iterations = attr.ib(default=default_iterations)


def print_help():
    """Explain what this is and show the user possible commands so that they can do stuff"""
    help_msg = """
A sierpinski's carpet is a design that is derived from string or pattern replacement that is done over a number of
passes. An initial single value is replaced by a rule from a rule set where each rule defines how each value is
replaced. The number of rules is dependent on the number of possible values and output of each rule can only include
possible values.

This tool generates these carpets, but it also allows the user to set the rules and number of iterations. The user can
set a rule set and the number of iterations, but if not the default is a classic sierpinski carpet over 3 iterations. 
It also has a number of utilities as listed below.

    exit     (q)    - Exit the script.
    generate (g)    - Will generate and print out the textual represent the sierpinski pattern.
    help     (h)    - Display this message.
    print    (p)    - Reprints the last pattern that was generated.
    quit     (q,x)  - Exit the script.
    rules    (r)    - Go to the rule set up menu
    """

    print(help_msg)

def generate(cfg):
    """All this does is take the user's desired configs (or the default) and generate/print the associated sierpinski
    carpet."""
    iterations = input('Number of iterations: ')
    #initialze the output so that it is the first character of the "0" rule
    print('Generation started...')
    cfg.output = cfg.rules[0][0] + '\n'
    lcv = 0
    while lcv < int(iterations)-1:
        next_output = ''
        staged_line_1 = ''
        staged_line_2 = ''
        staged_line_3 = ''
        for focus in cfg.output:
            if focus == '\n':
                #End of line prep for next line
                next_output = next_output + '\n'.join([staged_line_1, staged_line_2, staged_line_3]) + '\n'
                staged_line_1 = ''
                staged_line_2 = ''
                staged_line_3 = ''
            else:
                try:
                    staged_line_1 = staged_line_1 + cfg.rules[int(focus)][0:3]
                    staged_line_2 = staged_line_2 + cfg.rules[int(focus)][3:6]
                    staged_line_3 = staged_line_3 + cfg.rules[int(focus)][6:9]
                except KeyError as e:
                    print('rule set error: Check your rules for {}. It doesn\'t have a rule'.format(e))
        cfg.output = next_output
        cfg.iterations = iterations
        lcv += 1
    print_output(cfg)
    return cfg

def rules(cfg):
    """front end function that guides rule set up."""
    rules_menu = """
********************************************************************************
Rules Setup
********************************************************************************
Options
  p   - print rules
  r   - reset defaults
  m   - modify rules
  x,q - return to the main menu
"""
    print(rules_menu)
    while 1==1:
        command = input('>')
        if command == 'p':
            print_rules(cfg)
            print('\n',rules_menu)
        elif command == 'r':
            cfg = reset_rules(cfg)
            print('\n',rules_menu)
        elif command == 'm':
            cfg = modify_rules(cfg)
            print('\n',rules_menu)
        elif command in ('x','q'):
            return cfg
        elif len(command) == 0:
            print('Type a command and then press enter')
        else:
            print('Invalid command', '\n\n', rules_menu)

def print_rules(cfg):
    print('Current rules')
    for rule_id, rule_value in cfg.rules.items():
        print(rule_id, ' > ', rule_value)

def reset_rules(cfg):
    cfg.rules = default_rules
    print('Rules reset to classic Sierpinski')
    return cfg

def modify_rules(cfg):
    print('Enter in new rules. q, x, or just enter to stop.')
    rule = 0
    new_rule_set = {}
    new_rule = input('{} > '.format(rule))
    while new_rule not in ('q', 'x', ''):
        if len(new_rule) == 9:
            new_rule_set[rule] = new_rule
            rule += 1
        else:
            print('Rule must be 9 characters long')
        new_rule = input('{} > '.format(rule))
    if len(new_rule_set) > 0:
        print('New rule set defined')
        print(new_rule_set)
        cfg.rules = new_rule_set
    else:
        print('New rule set is empty. Ignoring changes')
    return cfg

def print_output(cfg):
    """Convenience function that check if there is output to be printed and prints it."""
    if len(cfg.output) > 0:
        print(cfg.output)
    else:
        print('*'*40, 'Warning', '*'*40)
        print('A carpet hasn''t been generated yet. Use the "generate" command to generate one. See "help"')
        print('for more commands"')

def main():
    """Main function to drive the user interface"""
    cfg = SierpinskiConfig()
    print('*'*80)
    print('Sierpinski''s Carpet Generator')
    print('*'*80)
    while 1==1:
        command = input('>')
        if command in ('h', 'help'):
            print_help()
        elif command in ('q', 'x', 'quit', 'exit'):
            return 0
        elif command in ('g', 'generate'):
            cfg = generate(cfg)
        elif command in ('r', 'rules'):
            cfg = rules(cfg)
        elif command in ('p', 'print'):
            print_output(cfg)
        elif len(command) == 0:
            print('Type a command and then press enter')
        else:
            print('Invalid command. Try again or type "h" or "help" for an explanation.')

if __name__ == '__main__':
    main()


#Notes
#=====
# * Patterns are generated using string replacement
# * Need to generate 3 lines at once


sierpinski_0.3.py
import attr
from PIL import Image
#-------------------------------------------------------------------------------
# Sierpinski's carpet
# Version: 0.3
# Date: 2017-10-25
# By: D. Clark
# Description:
# --------------------------------------
# Simplistic utility to generate Sierpinski carpets. It's a command line driven
# script that will take basic commands to generate the patterns, and if desired,
# images of those patterns
#
# Todos:
# --------------------------------------
# * Testing
# * Logging
# * Investigate usage of matrices 
#-------------------------------------------------------------------------------
# Some sane defaults. Need a better place for this Using globals for now.
default_rules = {
    0:'000010000',
    1:'111111111'}
default_iterations = 3
colors = {'0': b'\xf3\x86\x30',
          '1': b'\x69\xd2\xe7',
          '2': b'\xe0\xe4\xcc',
          '3': b'\xa7\xdb\xd8',
          '4': b'\xfa\x69\x00',
        }
rules = None


@attr.s
class SierpinskiConfig(object):
    output = attr.ib(default='')
    rules = attr.ib(default=default_rules)
    iterations = attr.ib(default=default_iterations)


def print_help():
    """Explain what this is and show the user possible commands so that they can do stuff"""
    help_msg = """
A sierpinski's carpet is a design that is derived from string or pattern replacement that is done over a number of
passes. An initial single value is replaced by a rule from a rule set where each rule defines how each value is
replaced. The number of rules is dependent on the number of possible values and output of each rule can only include
possible values.

This tool generates these carpets, but it also allows the user to set the rules and number of iterations. The user can
set a rule set and the number of iterations, but if not the default is a classic sierpinski carpet over 3 iterations. 
It also has a number of utilities as listed below.

    exit     (q)    - Exit the script.
    generate (g)    - Will generate and print out the textual represent the sierpinski pattern.
    help     (h)    - Display this message.
    print    (p)    - Reprints the last pattern that was generated.
    quit     (q,x)  - Exit the script.
    rules    (r)    - Go to the rule set up menu
    """

    print(help_msg)

def generate(cfg):
    """All this does is take the user's desired configs (or the default) and generate/print the associated sierpinski
    carpet."""
    iterations = input('Number of iterations: ')
    #initialze the output so that it is the first character of the "0" rule
    print('Generation started...')
    cfg.output = cfg.rules[0][0] + '\n'
    lcv = 0
    while lcv < int(iterations)-1:
        next_output = ''
        staged_line_1 = ''
        staged_line_2 = ''
        staged_line_3 = ''
        for focus in cfg.output:
            if focus == '\n':
                #End of line prep for next line
                next_output = next_output + '\n'.join([staged_line_1, staged_line_2, staged_line_3]) + '\n'
                staged_line_1 = ''
                staged_line_2 = ''
                staged_line_3 = ''
            else:
                try:
                    staged_line_1 = staged_line_1 + cfg.rules[int(focus)][0:3]
                    staged_line_2 = staged_line_2 + cfg.rules[int(focus)][3:6]
                    staged_line_3 = staged_line_3 + cfg.rules[int(focus)][6:9]
                except KeyError as e:
                    print('rule set error: Check your rules for {}. It doesn\'t have a rule'.format(e))
        cfg.output = next_output
        cfg.iterations = iterations
        lcv += 1
    print_output(cfg)
    return cfg

def rules(cfg):
    """front end function that guides rule set up."""
    rules_menu = """
********************************************************************************
Rules Setup
********************************************************************************
Options
  p   - print rules
  r   - reset defaults
  m   - modify rules
  x,q - return to the main menu
"""
    print(rules_menu)
    while 1==1:
        command = input('>')
        if command == 'p':
            print_rules(cfg)
            print('\n',rules_menu)
        elif command == 'r':
            cfg = reset_rules(cfg)
            print('\n',rules_menu)
        elif command == 'm':
            cfg = modify_rules(cfg)
            print('\n',rules_menu)
        elif command in ('x','q'):
            return cfg
        elif len(command) == 0:
            print('Type a command and then press enter')
        else:
            print('Invalid command', '\n\n', rules_menu)

def print_rules(cfg):
    print('Current rules')
    for rule_id, rule_value in cfg.rules.items():
        print(rule_id, ' > ', rule_value)

def reset_rules(cfg):
    cfg.rules = default_rules
    print('Rules reset to classic Sierpinski')
    return cfg

def modify_rules(cfg):
    print('Enter in new rules. q, x, or just enter to stop.')
    rule = 0
    new_rule_set = {}
    new_rule = input('{} > '.format(rule))
    while new_rule not in ('q', 'x', ''):
        if len(new_rule) == 9:
            new_rule_set[rule] = new_rule
            rule += 1
        else:
            print('Rule must be 9 characters long')
        new_rule = input('{} > '.format(rule))
    if len(new_rule_set) > 0:
        print('New rule set defined')
        print(new_rule_set)
        cfg.rules = new_rule_set
    else:
        print('New rule set is empty. Ignoring changes')
    return cfg

def print_output(cfg):
    """Convenience function that check if there is output to be printed and prints it."""
    if len(cfg.output) > 0:
        print(cfg.output)
    else:
        print('*'*40, 'Warning', '*'*40)
        print('A carpet hasn''t been generated yet. Use the "generate" command to generate one. See "help"')
        print('for more commands"')

def chunker(s):
    """Spilt s string into 8 long chunks and convert to ints"""
    for i in range(0, len(s), 8):
        yield int(s[i:i+8],2)

def generate_image(cfg):
    size = 3 ** (int(cfg.iterations)-1)
    color_count = len(cfg.rules)

    #For two rules we need to padd and chunk
    if color_count == 2:
        padding = '0'*(8-size%8)
        stripped_data = cfg.output.translate({ord(c): padding for c in '\n'})
        refined_data = bytes(list(chunker(stripped_data)))
        mode = '1'

    if color_count > 2:
        stripped_data = cfg.output.translate({ord(c): None for c in '\n'})
        refined_data = b''.join([colors[n] for n in stripped_data])
        mode = 'RGB'

    image = Image.frombytes(mode, (size, size), refined_data)
    image.save('test.bmp')

def main():
    """Main function to drive the user interface"""
    cfg = SierpinskiConfig()
    print('*'*80)
    print('Sierpinski''s Carpet Generator')
    print('*'*80)
    while 1==1:
        command = input('>')
        if command in ('h', 'help'):
            print_help()
        elif command in ('q', 'x', 'quit', 'exit'):
            return 0
        elif command in ('g', 'generate'):
            cfg = generate(cfg)
        elif command in ('r', 'rules'):
            cfg = rules(cfg)
        elif command in ('p', 'print'):
            print_output(cfg)
        elif command in ('i', 'image'):
            generate_image(cfg)
        elif len(command) == 0:
            print('Type a command and then press enter')
        else:
            print('Invalid command. Try again or type "h" or "help" for an explanation.')

if __name__ == '__main__':
    main()


#Notes
#=====
# * Patterns are generated using string replacement
# * Need to generate 3 lines at once
